cilium/CFP-33716-distributed-packet-tracing.md

```diff
@@ -0,0 +1,424 @@
+    check_and_store_ip_trace_id(ctx, ENABLE_PACKET_IP_TRACING);
+    // Do parsing and storing of ip_trace_id
+#endif
+```
+```c
+// Function to parse trace ID and store it if valid
+static __always_inline void check_and_store_ip_trace_id(struct __ctx_buff *ctx, __u8 ip_opt_type_value) {
+    __s64 trace_id;
+    trace_id = trace_id_from_ctx(ctx, ip_opt_type_value); // do parsing
+
+    // check if the trace ID is valid
+    if (trace_id > 0) {
+        bpf_trace_id_set(trace_id);
+    } else {
+        bpf_trace_id_set(0);
+    }
+}
+```
+At ingress, the function `trace_id_from_ctx` is called to parse the IP options from incoming packets. If the packet is valid, it proceeds to parse the IP options to extract the trace ID.
+
+We loop over the IP options to extract the trace ID. This loop is unrolled to a maximum of three iterations to minimize performance impact and meet BPF requirements. The maximum of three iterations accounts for any combination of trace ID, user-provided options, and either egress NAT or DSR options. If we read an IP option from the trace type which we configured (see Section 2), then this is stored as the `ip_trace_id`.
+
+Several edge cases are considered during parsing:
+- **No IP Options:** If the packet header indicates there are no options, parsing stops early, avoiding unnecessary processing.
+- **Invalid `ip_trace_id`:** invalid IDs are handled by setting the trace ID to 0.
+
+## Section 4: `ip_trace_id` storing
+
+Creating a per-CPU array-map allows for storage of each packet’s `ip_trace_id` as it is processed. When the feature gate is disabled (see Section 2: Feature enablement), the map is not rendered, avoiding space inefficiencies.
+
+```c
+#ifdef ENABLE_PACKET_IP_TRACING 
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+    __uint(max_entries, 1);
+    __type(key, __u32); // only one key
+    __type(value, __u64); // ip_trace_id type
+    __uint(pinning, LIBBPF_PIN_BY_NAME);
+} ip_trace_id_map __section_maps_btf;
+...
+// remaining declarations 
+#endif
+```
+
+**Pros**
+- Non-competitive storage.
+
+**Considerations**
+- Must ensure that data is cleared correctly.
+- This is universal storage and not specifically stored in the SKB struct, so it must be cleared when a new packet is processed.
+- At an ingress point, if there is no trace ID stored in a packet, the stored `ip_trace_id` is cleared from the array-map.
+
+## Section 5: Event integration
+
+Include ID as event field for trace and drop structs and the flow message (for Hubble integration):
+
+```c
+struct trace_notify {
+    // Existing struct members
+    ...
+    __u64      ip_trace_id; // new trace ID field
+    // Remaining struct members
+    ...
+};
+```
+```c
+struct drop_notify {
+    // Existing struct members
+    ...
+    __u64      ip_trace_id; // new trace ID field
+    // Remaining struct members
+    ...
+};
+```
+
+```go
+type Flow struct {
+    // Existing struct members
+    ...
+    IpTraceID uint64 
```

Comment:
A few thoughts:
- We already have a `uuid` for the Flow, and I wonder if we should consider using that.
- Since this is part of the IP packet, perhaps it should be part of the `IP` field instead?

Reply:
Creating a message type for the IpTraceID makes a lot of sense; having the IP option recorded would be helpful additional information. These traces are used to correlate flows, so I think creating a new message type and having a field in the Flow struct is an appropriate way to structure this. This has been updated in the new commits. If there are any additional thoughts, please let me know. If this answered your comment, please resolve the comment. Thank you!